README.md
<img width="100" src="https://user-images.githubusercontent.com/11070996/147922657-3c079672-edbd-4993-a645-f71a2739b18c.png#gh-dark-mode-only"/>
<img width="100" src="https://user-images.githubusercontent.com/11070996/147922660-890e2d96-26ee-4358-afc4-8421e9a05d5d.png#gh-light-mode-only"/>


# Chikamichi - Quickly find a page
<a href="https://chrome.google.com/webstore/detail/chikamichi-quickly-find-a/gkhobepjbiepngbeikhbpnfgjcjgmgha"><img alt="Chrome web store version" src="https://img.shields.io/chrome-web-store/v/gkhobepjbiepngbeikhbpnfgjcjgmgha.svg"></a>
<a href="https://chrome.google.com/webstore/detail/chikamichi-quickly-find-a/gkhobepjbiepngbeikhbpnfgjcjgmgha"><img alt="Chrome web store rating" src="https://img.shields.io/chrome-web-store/stars/gkhobepjbiepngbeikhbpnfgjcjgmgha.svg"></a>
<a href="https://chrome.google.com/webstore/detail/chikamichi-quickly-find-a/gkhobepjbiepngbeikhbpnfgjcjgmgha"><img alt="Chrome Web Store users" src="https://img.shields.io/chrome-web-store/users/gkhobepjbiepngbeikhbpnfgjcjgmgha"></a>
[![CI:UT](https://github.com/kawamataryo/chikamichi/actions/workflows/cypress-test.yaml/badge.svg)](https://github.com/kawamataryo/chikamcichi/actions/workflows/cypress-test.yaml)
[![CI:E2E](https://github.com/kawamataryo/chikamichi/actions/workflows/test.yaml/badge.svg)](https://github.com/kawamataryo/chikamcichi/actions/workflows/test.yaml)

<a href="https://www.producthunt.com/posts/chikamichi?utm_source=badge-featured&utm_medium=badge&utm_souce=badge-chikamichi" target="_blank"><img src="https://api.producthunt.com/widgets/embed-image/v1/featured.svg?post_id=328833&theme=light" alt="Chikamichi - Chrome extension that enables fuzzy search for anything | Product Hunt" style="width: 250px; height: 54px;" width="250" height="54" /></a>


Chikamichi is an extension that makes it very easy to navigate pages in Chrome or Firefox. You can search and navigate across browser history, bookmarks, and tabs.
Simple shortcuts and a simple interface make it easy to use.  

Inspired by [Sidekick](https://www.meetsidekick.com/) search dialog.  
The Japanese meaning of `chikamichi` is a shorter way.
  
## ✨ Features

- ⚡️ Fuzzy search your browsing history and bookmark and tab. powered by [Fuse.js](https://fusejs.io/)
- 🔐 All processing is done within the browser. No history data will be sent to the any server.

## 🎬 Demo

https://user-images.githubusercontent.com/11070996/151462764-4c196ea8-e5d0-4190-be9b-e5d79bf454ab.mp4

## 📦 Install

Please install from:
- [Chrome Web Store](https://chrome.google.com/webstore/detail/chikamichi/gkhobepjbiepngbeikhbpnfgjcjgmgha)
- [Firefox Add-Ons Repository (AMO)](https://addons.mozilla.org/firefox/addon/chikamichi/)
- [Edge Add-ons](https://microsoftedge.microsoft.com/addons/detail/chikamichi-quickly-find/kgbibnihcjkbcjbngneigamkphnoipli)

## 💻 Usage

### Shortcuts

| shortcut                              | action                                   |
|---------------------------------------|------------------------------------------|
| `Alt + k`                             | Open search dialog                       |
| `↓` or `↑` (`Ctrl + n` or `Ctrl + p`) | Select history                           |
| `Enter`                               | Open the selected url            |
| `Ctrl + Enter`                        | Open the selected url in new tab |
| `Ctrl + f`                            | Add favorite |
| `Ctrl + c`                            | Copy URL to the clipboard |

### Search commands

| commands | action                |
|----------|-----------------------|
| `/h`     | Search only histories |
| `/b`     | Search only bookmarks |
| `/t`     | Search only tabs      |

### Favorite
Items with a star to the right of the search item are registered as favorites. Items registered as favorites will be displayed in the initial view. However, if a search prefix has been set, the search prefix will take precedence.

## 👨‍💻 Contributing
Contributions are welcome 🎉 We accept contributions via Pull Requests.

See [this guide](https://github.com/kawamataryo/chikamichi/blob/main/CONTRIBUTING.md) on how to make a contribution.

## 💕 Thanks
This extension uses the following library.

* [Viteese-webext](https://github.com/antfu/vitesse-webext)
* [Fuse.js](https://fusejs.io/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 3.3.1
### design
- Hide scrollbar

## 3.3.0
### feature
- Add last viewed date to search order determination #649

## 3.2.0
### feature
- Add config for default link clicking action  #639

## 3.1.0
### performance
- Move fuse search to the service worker #638

## 2.17.0
### feature
- Add a feature to copy to the clipboard

## 2.16.0
### refactor
- optimize Fuse.js option by @yamash723

## 2.15.1
### chore
- adjust debounced time

## 2.14.3
### chore
- fix Fixed maximum number of history retrieval

## 2.14.2
### bugfix
- Fix favicons to be more visible

## 2.14.1
### bugfix
- fixed degrading of url display

## 2.14.0
### feat
- set url if bookmark title is empty by @yktakaha4

### chore
- update dependencies

## 2.13.3
### refactor
- move search logic to useSearch
- change highlight logic to use fuse matches

## 2.13.2
### chore
- fix favorite logic

## 2.13.1
### fix
- fix history search settings

## 2.13.0
### feat
- improve search performance.
- change highlight logic

## 2.12.0
### feat
- improve search performance.
- show all parent folder name

## 2.11.2
### feat
- open new tab with cmd + enter

### clean
- add favorite click event test

## 2.11.1
### clean
- update dependencies
- add cypress test
- fix type error
- lint and format code

## 2.11.0
### feature
- add favorite shortcut

## 2.10.0
### feature
- add favorite items

## 2.9.0
### feature
- expand the number of search displays #26

## 2.8.0
### feature
- remove same title histories
- adding submit github action by @louisgv #24

## 2.7.2
### bugfix
- fixed overflow bug

## 2.7.1
### bugfix
- fixed to not show top-level bookmark folder name

## 2.7.0
### feature
- search bookmarks by parent folder name

## 2.6.3
### design
- fixed the layout of search results #21

## 2.6.2
### chore
- fix overflow-scroll. #19 thanks @Ademking

## 2.6.1
### chore
- add autocomplete off. #17 thanks @Ademking

## 2.6.0
### feature
- add theme setting

## 2.5.1
### bugfix
- fixed a problem with ctrl+enter not working on Windows #14

## 2.5.0
### feature
- use browser's search engine instead of Google. #12 thanks  @Merlin04

### bugfix
- fix overflow bug. #11 thanks @Merlin04

## 2.4.0
### feature
- add setting page and default search prefix

## 2.3.2
### chore
- fix button color

## 2.3.1
### fix
- fix selected color

## 2.3.0
### feature
- Add side menu and info page

## 2.2.0
### fix
- Improve search performance.
  - Enabled Fuse ExtendedSearch option

## 2.1.0
### fix
- Improve search performance.

## 2.0.0
### breaking change
- Stop injecting dom in ContentScript, and show form in popup.

## 1.2.0
### feat
- add highlight matches from [@turara](https://github.com/turara)
### internal
- add vitest


src/constants.ts
import type Fuse from "fuse.js";

export const SEARCH_TARGET_REGEX = {
  HISTORY: /^\/h\s(.*)/,
  BOOKMARK: /^\/b\s(.*)/,
  TAB: /^\/t\s(.*)/,
  EITHER: /^\/[hbt]\s(.*)/,
} as const;

export const SEARCH_ITEM_TYPE = {
  HISTORY: "history",
  BOOKMARK: "bookmark",
  TAB: "tab",
} as const;

export const FUSE_OPTIONS: Fuse.IFuseOptions<any> = {
  useExtendedSearch: true,
  distance: 300,
  minMatchCharLength: 2,
  includeScore: true,
  includeMatches: true,
  shouldSort: true,
  keys: [
    {
      name: "title",
      weight: 0.4,
    },
    {
      name: "folderName",
      weight: 0.3,
    },
    {
      name: "url",
      weight: 0.2,
    },
    {
      name: "searchTerm",
      weight: 0.1,
    },
  ],
  threshold: 0.4,
};

export const PAGES = {
  SEARCH: "search",
  INFO: "info",
  SETTING: "setting",
} as const;

export const SEARCH_PREFIX = {
  HISTORY: "/h ",
  BOOKMARK: "/b ",
  TAB: "/t ",
};

export const THEME = {
  AUTO: "auto",
  DARK: "dark",
  LIGHT: "light",
} as const;

export const SEARCH_ICON_DATA_URL_LIGHT =
  'data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" role="img" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" preserveAspectRatio="xMidYMid meet" viewBox="0 0 24 24"><path d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"></path></svg>';

export const SEARCH_ICON_DATA_URL_DARK =
  'data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" role="img" fill="none" stroke="lightgrey" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" preserveAspectRatio="xMidYMid meet" viewBox="0 0 24 24"><path d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"></path></svg>';

export const BADGE_TEXT = {
  ADD_FAVORITE: "Add to favorite",
  REMOVE_FAVORITE: "Removed from favorite ",
  COPY: "Copied to clipboard",
} as const;
